File: blueprints/search/pkg/engine/local/engines/youtube.go
package engines

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

// YouTube implements YouTube video search (no API key required).
type YouTube struct {
	*BaseEngine
	timeRangeMap map[TimeRange]string
}

// NewYouTube creates a new YouTube engine.
func NewYouTube() *YouTube {
	y := &YouTube{
		BaseEngine: NewBaseEngine("youtube", "yt", []Category{CategoryVideos, CategoryMusic}),
		timeRangeMap: map[TimeRange]string{
			TimeRangeDay:   "Ag",
			TimeRangeWeek:  "Aw",
			TimeRangeMonth: "BA",
			TimeRangeYear:  "BQ",
		},
	}

	y.SetPaging(true).
		SetTimeRangeSupport(true).
		SetTimeout(5 * time.Second).
		SetAbout(EngineAbout{
			Website:    "https://www.youtube.com",
			WikidataID: "Q866",
			Results:    "HTML+JSON",
		})

	return y
}

func (y *YouTube) Request(ctx context.Context, query string, params *RequestParams) error {
	queryParams := url.Values{}
	queryParams.Set("search_query", query)

	// Time range filter
	if params.TimeRange != "" {
		if sp, ok := y.timeRangeMap[params.TimeRange]; ok {
			queryParams.Set("sp", "EgIIA"+sp+"%3D%3D")
		}
	}

	params.URL = "https://www.youtube.com/results?" + queryParams.Encode()
	params.Headers.Set("Accept", "text/html")
	params.Headers.Set("Accept-Language", "en-US,en;q=0.9")
	params.Cookies = append(params.Cookies, &http.Cookie{Name: "CONSENT", Value: "YES+"})

	return nil
}

func (y *YouTube) Response(ctx context.Context, resp *http.Response, params *RequestParams) (*EngineResults, error) {
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	results := NewEngineResults()

	// Extract ytInitialData JSON from HTML
	re := regexp.MustCompile(`var ytInitialData = ({.+?});</script>`)
	matches := re.FindSubmatch(body)
	if len(matches) < 2 {
		return results, nil
	}

	var data ytInitialData
	if err := json.Unmarshal(matches[1], &data); err != nil {
		return results, nil
	}

	// Navigate to video results
	for _, content := range data.Contents.TwoColumnSearchResultsRenderer.PrimaryContents.SectionListRenderer.Contents {
		for _, item := range content.ItemSectionRenderer.Contents {
			if item.VideoRenderer.VideoID != "" {
				vr := item.VideoRenderer

				// Build title from runs
				var title strings.Builder
				for _, run := range vr.Title.Runs {
					title.WriteString(run.Text)
				}

				// Build description from runs
				var desc strings.Builder
				for _, run := range vr.DescriptionSnippet.Runs {
					desc.WriteString(run.Text)
				}

				result := Result{
					URL:       fmt.Sprintf("https://www.youtube.com/watch?v=%s", vr.VideoID),
					Title:     title.String(),
					Content:   desc.String(),
					Template:  "videos",
					Duration:  vr.LengthText.SimpleText,
					EmbedURL:  fmt.Sprintf("https://www.youtube-nocookie.com/embed/%s", vr.VideoID),
					IFrameSrc: fmt.Sprintf("https://www.youtube-nocookie.com/embed/%s", vr.VideoID),
				}

				// Get author/channel
				if len(vr.OwnerText.Runs) > 0 {
					result.Artist = vr.OwnerText.Runs[0].Text
				}

				// Get thumbnail (last one is usually highest quality)
				if len(vr.Thumbnail.Thumbnails) > 0 {
					result.ThumbnailURL = vr.Thumbnail.Thumbnails[len(vr.Thumbnail.Thumbnails)-1].URL
				}

				result.ParsedURL, _ = url.Parse(result.URL)
				results.Add(result)
			}
		}
	}

	return results, nil
}

// YouTube JSON structures
type ytInitialData struct {
	Contents struct {
		TwoColumnSearchResultsRenderer struct {
			PrimaryContents struct {
				SectionListRenderer struct {
					Contents []struct {
						ItemSectionRenderer struct {
							Contents []struct {
								VideoRenderer ytVideoRenderer `json:"videoRenderer"`
							} `json:"contents"`
						} `json:"itemSectionRenderer"`
					} `json:"contents"`
				} `json:"sectionListRenderer"`
			} `json:"primaryContents"`
		} `json:"twoColumnSearchResultsRenderer"`
	} `json:"contents"`
}

type ytVideoRenderer struct {
	VideoID   string `json:"videoId"`
	Thumbnail struct {
		Thumbnails []struct {
			URL    string `json:"url"`
			Width  int    `json:"width"`
			Height int    `json:"height"`
		} `json:"thumbnails"`
	} `json:"thumbnail"`
	Title struct {
		Runs []struct {
			Text string `json:"text"`
		} `json:"runs"`
	} `json:"title"`
	DescriptionSnippet struct {
		Runs []struct {
			Text string `json:"text"`
		} `json:"runs"`
	} `json:"descriptionSnippet"`
	LengthText struct {
		SimpleText string `json:"simpleText"`
	} `json:"lengthText"`
	OwnerText struct {
		Runs []struct {
			Text string `json:"text"`
		} `json:"runs"`
	} `json:"ownerText"`
	ViewCountText struct {
		SimpleText string `json:"simpleText"`
	} `json:"viewCountText"`
	PublishedTimeText struct {
		SimpleText string `json:"simpleText"`
	} `json:"publishedTimeText"`
}

File: blueprints/search/app/worker/src/engines/youtube.ts
/**
 * YouTube Search Engine adapter.
 * Ported from Go: pkg/engine/local/engines/youtube.go
 *
 * Scrapes ytInitialData JSON from YouTube search results page.
 * No API key required.
 */

import type {
  OnlineEngine,
  EngineParams,
  RequestConfig,
  EngineResults,
  Category,
} from './engine';
import { newEngineResults } from './engine';

// Time range encoding in sp parameter (YouTube's filter encoding)
const timeRangeMap: Record<string, string> = {
  day: 'Ag',
  week: 'Aw',
  month: 'BA',
  year: 'BQ',
};

// ========== YouTube JSON Types ==========

interface YtVideoRenderer {
  videoId?: string;
  thumbnail?: {
    thumbnails?: Array<{
      url?: string;
      width?: number;
      height?: number;
    }>;
  };
  title?: {
    runs?: Array<{ text?: string }>;
  };
  descriptionSnippet?: {
    runs?: Array<{ text?: string }>;
  };
  lengthText?: {
    simpleText?: string;
  };
  ownerText?: {
    runs?: Array<{ text?: string }>;
  };
  viewCountText?: {
    simpleText?: string;
  };
  publishedTimeText?: {
    simpleText?: string;
  };
}

interface YtInitialData {
  contents?: {
    twoColumnSearchResultsRenderer?: {
      primaryContents?: {
        sectionListRenderer?: {
          contents?: Array<{
            itemSectionRenderer?: {
              contents?: Array<{
                videoRenderer?: YtVideoRenderer;
              }>;
            };
          }>;
        };
      };
    };
  };
}

export class YouTubeEngine implements OnlineEngine {
  name = 'youtube';
  shortcut = 'yt';
  categories: Category[] = ['videos'];
  supportsPaging = true;
  maxPage = 5;
  timeout = 5_000;
  weight = 1.0;
  disabled = false;

  buildRequest(query: string, params: EngineParams): RequestConfig {
    const searchParams = new URLSearchParams();
    searchParams.set('search_query', query);

    // Time range filter encoded in sp parameter
    if (params.timeRange && timeRangeMap[params.timeRange]) {
      const sp = `EgIIA${timeRangeMap[params.timeRange]}%3D%3D`;
      searchParams.set('sp', sp);
    }

    return {
      url: `https://www.youtube.com/results?${searchParams.toString()}`,
      method: 'GET',
      headers: {
        Accept: 'text/html',
        'Accept-Language': 'en-US,en;q=0.9',
      },
      cookies: ['CONSENT=YES+'],
    };
  }

  parseResponse(body: string, _params: EngineParams): EngineResults {
    const results = newEngineResults();

    // Extract ytInitialData JSON from HTML
    const match = body.match(
      /var ytInitialData = ({.+?});<\/script>/
    );
    if (!match) return results;

    let data: YtInitialData;
    try {
      data = JSON.parse(match[1]);
    } catch {
      return results;
    }

    // Navigate to video results
    const sectionContents =
      data.contents?.twoColumnSearchResultsRenderer?.primaryContents
        ?.sectionListRenderer?.contents;

    if (!sectionContents) return results;

    for (const section of sectionContents) {
      const items = section.itemSectionRenderer?.contents;
      if (!items) continue;

      for (const item of items) {
        const vr = item.videoRenderer;
        if (!vr?.videoId) continue;

        // Build title from runs
        let title = '';
        if (vr.title?.runs) {
          title = vr.title.runs.map((r) => r.text || '').join('');
        }

        // Build description from runs
        let description = '';
        if (vr.descriptionSnippet?.runs) {
          description = vr.descriptionSnippet.runs
            .map((r) => r.text || '')
            .join('');
        }

        // Get channel name
        let channel = '';
        if (vr.ownerText?.runs && vr.ownerText.runs.length > 0) {
          channel = vr.ownerText.runs[0].text || '';
        }

        // Get thumbnail (last one is usually highest quality)
        let thumbnailUrl = '';
        const thumbnails = vr.thumbnail?.thumbnails;
        if (thumbnails && thumbnails.length > 0) {
          thumbnailUrl = thumbnails[thumbnails.length - 1].url || '';
        }

        // Duration
        const duration = vr.lengthText?.simpleText || '';

        // View count
        let views = 0;
        const viewText = vr.viewCountText?.simpleText || '';
        const viewMatch = viewText.match(/([\d,]+)\s*view/i);
        if (viewMatch) {
          views = parseInt(viewMatch[1].replace(/,/g, ''), 10) || 0;
        }

        const videoUrl = `https://www.youtube.com/watch?v=${vr.videoId}`;
        const embedUrl = `https://www.youtube-nocookie.com/embed/${vr.videoId}`;

        results.results.push({
          url: videoUrl,
          title,
          content: description,
          engine: this.name,
          score: this.weight,
          category: 'videos',
          template: 'videos',
          duration,
          embedUrl,
          thumbnailUrl,
          channel,
          views,
        });
      }
    }

    return results;
  }
}
